fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/KeyExpressionVisitor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * KeyExpressionVisitor.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import com.apple.foundationdb.record.metadata.expressions.EmptyKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.FieldKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpressionWithValue;
+import com.apple.foundationdb.record.metadata.expressions.KeyWithValueExpression;
+import com.apple.foundationdb.record.metadata.expressions.NestingKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.ThenKeyExpression;
+
+import javax.annotation.Nonnull;
+
+/**
+ * An interface so provide a state-based visitor pattern that can traverse a tree of {@link KeyExpression}s.
+ * @param <S> the type of the state object which depends on the specific implementation
+ * @param <R> the type of the result object which is returned by all visitation methods
+ */
+public interface KeyExpressionVisitor<S extends KeyExpressionVisitor.State, R extends KeyExpressionVisitor.Result> {
+
+    /**
+     * Method to return the (immutable) current state.
+     * @return the current state of type {@code S}
+     */
+    S getCurrentState();
+
+    /**
+     * Default method that is called on unknown sub classes of {@link KeyExpression}. That makes it possible to
+     * add a {@code visitor.visitExpression(this)} regardless whether the visitor defines an actual specific override
+     * for the sub class of {@code this} or not.
```

Comment:
I know we have other instances of this same pattern, but I am not sure everyone is happy with them. Specifically, this makes it harder to tell just by looking whether a specific case is supported specially or not. For non-`this` cases, there are also problems where otherwise innocent widening loses the right method.

Reply:
It's true. Will leave this pattern in for now unless you object. I would imagine fielding with this in the near future when we add more sorts of key expressions.

Reply:
✅